File: next/kmp/pureHttp/src/commonMain/kotlin/org/dweb_browser/pure/http/PureMethod.kt
package org.dweb_browser.pure.http

import io.ktor.http.HttpMethod
import kotlinx.serialization.Serializable
import org.dweb_browser.helper.StringEnumSerializer

object PureMethodSerializer :
  StringEnumSerializer<PureMethod>("MatchMode", PureMethod.ALL_VALUES, { method })

/**
 * Ipc 使用的 Method
 */
@Serializable(PureMethodSerializer::class)
enum class PureMethod(val method: String, val ktorMethod: HttpMethod) {
  GET("GET", HttpMethod.Get),
  POST("POST", HttpMethod.Post),
  PUT("PUT", HttpMethod.Put),
  DELETE("DELETE", HttpMethod.Delete),
  OPTIONS("OPTIONS", HttpMethod.Options),
  TRACE("TRACE", HttpMethod.Post),
  CONNECT("CONNECT", HttpMethod.Post),
  PATCH("PATCH", HttpMethod.Patch),
  PURGE("PURGE", HttpMethod.Post),
  HEAD("HEAD", HttpMethod.Head),
  ;

  companion object {
    val ALL_VALUES = PureMethod.entries.associateBy { it.method }
    fun from(ktorMethod: HttpMethod) = from(ktorMethod.value)
    fun from(method: String) = ALL_VALUES[method.uppercase()] ?: GET
  }
}

File: toolkit/dweb-core/packages/js-core/src/ipc/helper/PureMethod.ts
export const enum PURE_METHOD {
  GET = "GET",
  POST = "POST",
  PUT = "PUT",
  DELETE = "DELETE",
  OPTIONS = "OPTIONS",
  TRACE = "TRACE",
  PATCH = "PATCH",
  PURGE = "PURGE",
  HEAD = "HEAD",
}
export const toPureMethod = (method?: string) => {
  if (method == null) {
    return PURE_METHOD.GET;
  }

  switch (method.toUpperCase()) {
    case PURE_METHOD.GET: {
      return PURE_METHOD.GET;
    }
    case PURE_METHOD.POST: {
      return PURE_METHOD.POST;
    }
    case PURE_METHOD.PUT: {
      return PURE_METHOD.PUT;
    }
    case PURE_METHOD.DELETE: {
      return PURE_METHOD.DELETE;
    }
    case PURE_METHOD.OPTIONS: {
      return PURE_METHOD.OPTIONS;
    }
    case PURE_METHOD.TRACE: {
      return PURE_METHOD.TRACE;
    }
    case PURE_METHOD.PATCH: {
      return PURE_METHOD.PATCH;
    }
    case PURE_METHOD.PURGE: {
      return PURE_METHOD.PURGE;
    }
    case PURE_METHOD.HEAD: {
      return PURE_METHOD.HEAD;
    }
  }
  throw new Error(`invalid method: ${method}`);
};
